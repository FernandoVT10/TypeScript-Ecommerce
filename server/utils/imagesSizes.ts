export const PRODUCTS_IMAGE_SIZES = [
    { label: "large", width: 1920, height: 1080 },
    { label: "medium", width: 1024, height: 768 },
    { label: "thumb", width: 250, height: 250 }
];

export const CAROUSEL_IMAGE_SIZES = [
    { label: "large", width: 1920, height: 1080 },
    { label: "medium", width: 1024, height: 575 },
    { label: "small", width: 500, height: 280 }
];
