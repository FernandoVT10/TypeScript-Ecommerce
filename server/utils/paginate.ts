export interface IPagination {
    hasPrevPage: boolean,
    prevPage: number,
    hasNextPage: boolean,
    nextPage: number,
    pages: {
        active: boolean,
        pageNumber: number
    }[]
}

export default function paginate(totalPages: number, cursor: number): IPagination {
    let remainingLinks = 6;
    let minPage = 0, maxPage = 0;

    if(cursor + 3 > totalPages) {
        remainingLinks -= totalPages - cursor;
        maxPage = totalPages;
    } else {
        remainingLinks -= 3;
        
        if(cursor < 4) {
            if(totalPages > 7) {
                maxPage = 7;
            } else {
                maxPage = totalPages;
            }
        } else {
            maxPage = cursor + 3;
        }
    }

    if(cursor - remainingLinks > 0) {
        minPage = cursor - remainingLinks;
    } else {
        minPage = 1;
    }

    const result: IPagination = {
        hasPrevPage: false,
        prevPage: null,
        hasNextPage: false,
        nextPage: null,
        pages: []
    };

    if(cursor > 1) {
        result.hasPrevPage = true;
        result.prevPage = cursor - 1;
    }

    if(cursor < totalPages) {
        result.hasNextPage = true;
        result.nextPage = cursor + 1;
    }

    for(let i = minPage; i <= maxPage; i++) {
        if(i === cursor) {
            result.pages.push({
                active: true,
                pageNumber: i
            });
        } else {
            result.pages.push({
                active: false,
                pageNumber: i
            });
        }
    }

    return result;
}