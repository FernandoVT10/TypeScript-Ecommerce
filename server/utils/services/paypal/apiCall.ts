import fetch from "node-fetch";

import { PAYPAL_CLIENT_ID, PAYPAL_CLIENT_SECRET } from "../../../config";

const PAYPAL_API = "https://api.sandbox.paypal.com";

export default async (url: string, body: object = {}) => {
    const authorization = `${PAYPAL_CLIENT_ID}:${PAYPAL_CLIENT_SECRET}`;
    const base64Authorization = Buffer.from(authorization).toString("base64");

    const res = await fetch(PAYPAL_API + url, {
	method: "POST",
	headers: {
	    "Content-Type": "application/json",
	    "Authorization": `Basic ${base64Authorization}`
	},
	body: JSON.stringify(body)
    });
    
    return await res.json();
}
