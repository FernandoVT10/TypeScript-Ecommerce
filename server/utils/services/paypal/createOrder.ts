import apiCall from "./apiCall";

const CURRENCY = "USD";

export interface IOrderItem {
    title: string,
    price: number,
    quantity: number
}

export default async (items: IOrderItem[]) => {
    const totalPrice = items.reduce((acc, item) => {
	return acc + item.quantity * item.price;
    }, 0);

    const orderItems = items.map(item => {
	return {
	    name: item.title,
	    unit_amount: {
		currency_code: CURRENCY,
		value: item.price.toFixed(2)
	    },
	    quantity: item.quantity.toString()
	};
    });

    try {
	const data = await apiCall("/v2/checkout/orders", {
	    intent: "CAPTURE",
	    purchase_units: [
		{
		    amount: {
			currency_code: CURRENCY,
			value: totalPrice.toFixed(2),
			breakdown: {
			    item_total: {
				currency_code: CURRENCY,
				value: totalPrice.toFixed(2)
			    }
			}
		    },
		    items: orderItems
		}
	    ]
	});

	if(!data.id) {
	    throw new Error("Payment couldn't be created"); 
	}

	return data.id;
    } catch(err) {
       throw new Error(err.message); 
    }
}
