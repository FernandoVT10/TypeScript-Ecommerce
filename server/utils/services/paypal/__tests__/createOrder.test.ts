import { mocked } from "ts-jest/utils";

import createOrder from "../createOrder";

import apiCall from "../apiCall";

jest.mock("../apiCall");

const mockedAPICall = mocked(apiCall);

const ORDER_ITEMS_MOCK = [
    {
	title: "test 1",
	price: 50,
	quantity: 2
    },
    {
	title: "test 2",
	price: 2.5,
	quantity: 4
    }
];

describe("Paypal Create Order service", () => {
    beforeEach(() => {
	mockedAPICall.mockReset();
	mockedAPICall.mockImplementation(() => Promise.resolve({ id: "testorderid" }));
    });

    it("should call apiCall and return the order id", async () => {
	const res = await createOrder(ORDER_ITEMS_MOCK);
	expect(res).toBe("testorderid");

	const call = mockedAPICall.mock.calls[0];
	
	expect(call[0]).toBe("/v2/checkout/orders");
	expect(call[1]).toMatchSnapshot();
    });

    it("should throw an error when the api call doesn't returns the order id", async () => {
	mockedAPICall.mockImplementation(() => Promise.resolve({}));

	try {
	    await createOrder(ORDER_ITEMS_MOCK);
	} catch (err) {
	    expect(err.message).toBe("Payment couldn't be created")
	}
    });
});
