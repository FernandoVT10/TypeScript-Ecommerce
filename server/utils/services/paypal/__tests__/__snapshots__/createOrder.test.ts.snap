// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paypal Create Order service should call apiCall and return the order id 1`] = `
Object {
  "intent": "CAPTURE",
  "purchase_units": Array [
    Object {
      "amount": Object {
        "breakdown": Object {
          "item_total": Object {
            "currency_code": "USD",
            "value": "110.00",
          },
        },
        "currency_code": "USD",
        "value": "110.00",
      },
      "items": Array [
        Object {
          "name": "test 1",
          "quantity": "2",
          "unit_amount": Object {
            "currency_code": "USD",
            "value": "50.00",
          },
        },
        Object {
          "name": "test 2",
          "quantity": "4",
          "unit_amount": Object {
            "currency_code": "USD",
            "value": "2.50",
          },
        },
      ],
    },
  ],
}
`;
